crates/bitwarden-crypto/src/opaque_ke/types.rs

```diff
@@ -0,0 +1,168 @@
+    pub ksf: Ksf,
+}
+
+#[derive(Serialize, Deserialize, Clone)]
+#[cfg_attr(feature = "wasm", derive(Tsify), tsify(into_wasm_abi, from_wasm_abi))]
+#[cfg_attr(feature = "uniffi", derive(uniffi::Enum))]
+pub enum OprfCs {
+    #[serde(rename = "ristretto255")]
+    Ristretto255,
+}
+
+#[derive(Serialize, Deserialize, Clone)]
+#[cfg_attr(feature = "wasm", derive(Tsify), tsify(into_wasm_abi, from_wasm_abi))]
+#[cfg_attr(feature = "uniffi", derive(uniffi::Enum))]
+pub enum KeGroup {
+    #[serde(rename = "ristretto255")]
+    Ristretto255,
+}
+
+#[derive(Serialize, Deserialize, Clone)]
+#[cfg_attr(feature = "wasm", derive(Tsify), tsify(into_wasm_abi, from_wasm_abi))]
+#[cfg_attr(feature = "uniffi", derive(uniffi::Enum))]
+pub enum KeyExchange {
+    #[serde(rename = "triple-dh")]
+    TripleDh,
+}
+
+#[derive(Serialize, Deserialize, Clone)]
+#[cfg_attr(feature = "wasm", derive(Tsify), tsify(into_wasm_abi, from_wasm_abi))]
+#[cfg_attr(feature = "uniffi", derive(uniffi::Enum))]
+pub enum Ksf {
+    #[serde(rename = "argon2id")]
+    Argon2id(u32, u32, u32),
+}
+
+#[derive(Serialize, Deserialize, Default, Clone)]
+#[cfg_attr(feature = "wasm", derive(Tsify), tsify(into_wasm_abi, from_wasm_abi))]
+pub struct Argon2Id {
+    /// Memory cost in kibibytes
+    pub memory_kib: u32,
+    /// Time cost in number of iterations
+    pub iterations: u32,
+    /// Number of threads that may be used
+    pub parallelism: u32,
+}
+
+impl opaque_ke::ksf::Ksf for Argon2Id {
+    fn hash<L: ArrayLength<u8>>(
+        &self,
+        input: GenericArray<u8, L>,
+    ) -> Result<GenericArray<u8, L>, InternalError> {
+        let mut output = GenericArray::default();
+        argon2::Argon2::new(
+            argon2::Algorithm::Argon2id,
+            argon2::Version::V0x13,
+            Params::new(
+                self.memory_kib,
+                self.iterations,
+                self.parallelism,
+                Some(output.len()),
+            )
+            .map_err(|_| InternalError::KsfError)?,
+        )
+        .hash_password_into(&input, &[0; argon2::RECOMMENDED_SALT_LEN], &mut output)
+        .map_err(|_| InternalError::KsfError)?;
+        Ok(output)
+    }
+}
+
+#[derive(Serialize, Deserialize)]
+#[cfg_attr(feature = "wasm", derive(Tsify), tsify(into_wasm_abi, from_wasm_abi))]
+pub struct RistrettoTripleDhArgonSuite {
+    pub ksf_params: Argon2Id,
+}
+
+impl CipherSuite for RistrettoTripleDhArgonSuite {
+    type OprfCs = opaque_ke::Ristretto255;
+    type KeGroup = opaque_ke::Ristretto255;
+    type KeyExchange = opaque_ke::key_exchange::tripledh::TripleDh;
+    type Ksf = Argon2Id;
+}
+
+#[derive(Serialize, Deserialize)]
+#[cfg_attr(feature = "wasm", derive(Tsify), tsify(into_wasm_abi, from_wasm_abi))]
+pub struct ClientRegistrationStartResult {
+    /// The client side, serialized state of the registration process.
+    /// MUST NOT BE SHARED WITH THE SERVER and must be cleared from memory after registration
+    /// finish.
+    pub state: ByteBuf,
```

Comment:
Just curious, what does `ByteBuf` become on the TS side? Is it special cased to become `ArrayBuffer`, `UInt8Array`, or does it copy the layout of this type? Is it actually easy to clear from memory?

Reply:
https://github.com/AmbientRun/tsify-next/blob/2e5e0a174fb4dd3b87bddec93286250001f57c46/tsify-macros/src/typescript.rs#L363

As far as I understand, this is the only way to get Uint8array directly in `tsify-next`. Otherwise you get number[], which has to be manually converted again.

Reply:
We generally encode these using b64 in other places. @dani-garcia not sure if that is more appropriate or if we revisit our practice. (I believe we b64 encode a few things for mobile at least)

Reply:
Yeah we've had complaints in the swift native apps when using bytes as they were hard to work with, something about equality not being implemented for the swift bytes type I think.

That said, using b64 is probably just as annoying and pushing the conversion work to the clients. I wonder if this just means we need to consider having our own `Bytes` type that does the right conversion for each platform.